### ISSUE TITLE
Data Streaming

### ISSUE THREAD

**Issue by dylanmcreynolds** [bluesky/tiled#437]
We recently had an interesting group conversation about streaming data. We discussed infrastructure that might be useful for a wide variety of use cases. I'm writing this in the Tiled repo because a) I want to put it somewhere and b) there may be a tie-in to Tiled. If that turns out not to be the case, I'll happily close this issue and put it somewhere else, but let's start here for now. 

Several facilities have expressed interest in streaming data from an instrument to a web browser.

There are a number of technologies that are potentially in use at beamlines for streaming data. Data can be small scalar or large arrays. Examples of these technologies include ZMQ, Kafka, PVA Stream. On the consumer side, it would be handy to provide streaming data in a way that provides a uniform transport protocol and well-known data structure. A central service that can take data from a variety of streaming protocols on one side and translate that to browser-friendly technologies would be a very useful tool to have. Additionally, providing authN/authZ infrastructure could allow this service to be used in remote access scenarios. While browsers are the easiest way to think about web-based technologies, such a service could provide non-browsers clients the ability to access streaming data over the web.

A number of technologies exist to provide server side events to the browser (server side events, websockets). WebSockets seems like a good place to start. It is widely adopted. It seems like firewalling issues that once existed for WebSockets have decreased over time. WebSockets supports two-way communication, which I could possibly see being useful in the far distant future.

Tiled already provides many things that I think we would want: 

- authentication infrastructure
- authorization infrastructure
- the facility to provide metadata to the client
- well-documented structures and a facility to convey them to client

I would like to get a discussion about the viability of creating such a service based on Tiled. Such a project could either be built into Tiled, or its own standalone repo and service that imports Tiled as a dependency.

**Comment by danielballan** [comment]
This is compelling. I think it is worth trying to build websockets support into core. We describe Tiled as offering:

* Slicing ("structured access" generally)
* Granular access control
* Transcoding
 
The last two would certainly apply to streaming data. Slicing applies too, though I suspect the semantics will be a different for streaming vs random access in some important ways.

Perhaps a given node in Tiled could support random access via HTTP and streaming access via web sockets (when applicable). This could include:

* Streaming documents from a Bluesky run, coming from Kafka
* Streaming updates from a CA or PVA channel

From a client POV the transition between streaming access and random access from storage at rest would be as smooth as can be: same authenticated session, same access control (authorization) rules, even a similar URL. Just a different protocol.

Months ago, I was initially resistant to this suggestion because I wanted to stay focused on getting the non-steaming use case rock solid. But now I am convinced.

**Comment by danielballan** [comment]
Saying a little more on what this could look like:

* An additional optional method, perhaps `stream()`, on Tiled adapters which is a generator returning items in the stream. Items would be small numpy arrays or DataFrames, corresponding to the structure_family of the Adapter. 
* A new route, maybe `/stream/{path}` with websockets, which calls the new method on the adapter, transcodes each item into the requested format, and yields it.

If our initial use case is streaming from a CA or PVA (EPICS) subscription, the Adapter should developed outside of Tiled core.

**Comment by danielballan** [comment]
Note that it is possible to develop this all out of Tiled core using the `include_routers` mechanism by which Adapters can add custom routes to the server. If we like it, seems likely we would move the /stream route into core.

**Comment by dylanmcreynolds** [comment]
This simple adapter interface will allow us to create an test an adapter while the development in Tiled starts up, which is great.

**Comment by danielballan** [comment]
Adding a diagram here that was sketched in a discussion about this with @whs92 and colleagues

![image](https://github.com/bluesky/tiled/assets/2279598/11a1ee70-6a82-4304-8ceb-7bf2ea0dda23)

**Comment by dylanmcreynolds** [comment]
I'm a little confused by something on the digram above. You have multiple routes between pymca and tiled. Is the two way http connection for something other than the ws connection?

**Comment by danielballan** [comment]
Yes. (And I am very happy for feedback on how to draw it better.)

```
GET /api/v1/metadata/{path}  # exists as soon as RunStart comes out
GET /api/v1/table/full/{path}  # poll for data during scan, or get all data after
ws /api/v1/stream/{path}  # stream data (while scan is live only)
```

**Comment by danielballan** [comment]
Picking this back up with @gwbischof and @hyperrealist. Summary of discussion:

Tiled is about data at rest, and it will _not_ grow Adapters that connect directly to streaming event sources. (At least not yet.) But it will grow the capability to:

1. Notify clients that new data is available (and how much)
2. Stream to clients data that it has just received (potentially transcoded, sliced/downsampled)

## Phases

### Phase 0: Dummy Prototype

Outside of Tiled, with no Tiled dependency, build a tiny starlette app, self-contained with all data in process memory. One route accepts data and appends it to a table (fixed schema, nothing complicated or flexible). This route may use websockets, SSE, or Server Push. _We should pick one_ but all are up for consideration. A second route notifies connected clients that new data is available. A third route _streams_ new data to connected clients over websockets.

Once this works, externalize the state to a PostgreSQL database. Use the PostgreSQL pubsub feature to support horizontally scaled servers. The client that submits new data may be connected to a different starlette server instance than the one which wants an update about the data.

Goals:
- Get to know technologies and how to debug them. Especially PostgreSQL pubsub and these various server-initiated update protcols.
- Validate that this works on horizontally scaled servers.
- Use this working minimal example to validate with stakeholders and collaborators that this generally satisfies our most important use cases.

### Phase 1: Integrate notifications with Tiled

New route:

```
/api/v1/notify/{path}
```

### Phase 2: Integrate streaming data with Tiled

New route, as previously suggested in earlier discussion:

```
/api/v1/stream/{path}  # stream data (while scan is live only)

**Comment by dylanmcreynolds** [comment]
I agree with this build-up approach. I would like to add a thought that came up on a call today.

For web browsers, websockets is a great technology. However, I have the impression that there are no really solid web socket clients in python. It would be nice to find a technology that is good for both browsers and for feeding analysis code.

Other technologies to possibly think about Server Side Events (SSE) and GrpahQL Subsction.

SSE is pretty simple. Unlike web sockets only provides communication in one direction, which is fine in this case.

GraphQL Subscriptions is more complicated, but I have no experience with it.

**Comment by dylanmcreynolds** [comment]
Regarding the above comment...maybe I'm wrong. This seems well-supported: https://github.com/python-websockets/websockets

**Comment by hyperrealist** [comment]
I have used above websockets library in the past when it was still not a matured project. Looking at their release notes, I think they have fixed most of the pain points.

My experience with GraphQL is only on the client side. But the obvious elephant in the room is that it would lead to _somewhat_ arbitrary compute requests to Tiled server, which I feel like we have avoided at every turn possible. Also, I have seen examples of seemingly innocuous queries that result in unauthorized data access. My impression is that GraphQL queries are hard to validate/sanitize.

**Comment by danielballan** [comment]
I agree. The latest rewrite of [python-websockets/websockets](https://github.com/python-websockets/websockets) uses the Sans I/O approach, same as h11 (and caproto). It looks well done and easy to integrate with Starlette on the server side and httpx on the client side.

Re-reading the [websockets RFC](https://datatracker.ietf.org/doc/html/rfc6455) for inspiration, it seems like most robust choice for both the "notify" and "stream" use cases. The key point is that you get a framing TCP-like protocol that can share a port with an HTTP server and thus be easy to deploy within existing facility infrastructure.

**Comment by danielballan** [comment]
Capturing some thoughts from a conversation with @gwbischof today:

**On picking up mid-stream** Borrowing [an idea from the _other_ Bluesky](https://atproto.com/specs/event-stream), the HTTP endpoint should accept a `cursor` query parameter that specify which item in the stream to start from. This is useful if a client is momentary disconnected and needs to catch up: it can tell the server where it left off and the server can replay recent history (within reason) before streaming the present. The section titled "Sequence Numbers" in the linked documentation has some useful thoughts on what to do in various possible cases.

**On push-backed notifications** When a request comes in to update a dataset, it is processed by one Tiled server. In a horizontally scaled deployment, how will the other servers find out about it and push to any connected websocket clients? When the Catalog database is backed by PostgreSQL, we can leverage PG's pubsub mechanism. A change can trigger a Python callback in each Tiled server, in turn notifying the websocket client. SQlite has no such mechanism. (I think it would be hard for them to add one robustly, given their embedded architecture and reliance of filesystem APIs.) For SQLite, I think our best available option is to degrade gracefully to using a polling loop, perhaps looking at the `time_updated` or other relevant state. For clients, the interaction would be the same; they would only notice a little more latency between the write and update when backed by SQLite.

**Comment by hyperrealist** [comment]
Prototype repo https://github.com/gwbischof/tiled-streaming-dev

**Comment by danielballan** [comment]
##  Notes on headers

- `Accept-Encoding` and `Content-Encoding` must have their standard usage / meaning because proxies may intervene.
- `Accept` and `Content-Type` should accurately describe the content because intervening layers or simple clients may try to parse the body. So, `image/tiff` should be TIFF and [`application/vnd.msgpack`](https://www.iana.org/assignments/media-types/application/vnd.msgpack) should be msgpack.

## Use cases

"Send me image stream as (unencapsulated) PNG or JPEG."
```
Accept: image/png, image/jpeg
Subprotocol: <plain>
```

where `<plain>` is a placeholder for some subprotocol we should name.

"Send me image stream as (unencapsulated) JSON arrays."
```
Accept: application/json
Subprotocol: <plain>
```

-> `[[1,2,3], [4,5,6]]`

"Send me image stream as JSON, encapsulated with info about frame number or whatever."
```
Accept: application/json
Subprotocol: <encapsulated>
```

-> `{"index": 3, "data": {"payload": [[1,2,3], [4,5,6]], "mimetype": "application/json"}}`

"Send me image as msgpack, encapsulated with info about frame number or whatever."
```
Accept: application/vnd.msgpack
Subprotocol: <encapsulated>
```

-> `{"index": 3, "data": {"payload": <C buffer>, "mimetype": "application/octet-stream"}}`

"Send me _table_ as msgpack, encapsulated with info about row numbers or whatever."
```
Accept: application/vnd.msgpack
Subprotocol: <encapsulated>
```

-> `{"index": 3, "data": {"payload': <Arrow buffer>, "mimetype": "application/vnd.apache.arrow...."}}`

**Comment by danielballan** [comment]
This comment is to document the evolving discussion with @dylanmcreynolds, @gwbischof, @hyperrealist, and others.

---

We now propose to involve Redis (or any Redis-like) in this architecture. Above, we had proposed using PostgreSQL's pub-sub feature (`PG NOTIFY`). But when we look at the all the requirements for streaming, Redis is the right tool for the job.

- Pub-sub is a core use case for Redis, whereas `PG NOTIFY` is a way to add a _little_ pubsub to a PG application. It is limited: for example, the body of the `PG NOTIFY` notification is text-only and small. We want pub-sub to be a core feature of Tiled, and we want to broadcast chunks of binary data to clients.
- The _way_ we store the data for streaming is different than the way we want to store it at rest. For streaming we want to be able to identify and efficiently access each unit of update, for rebroadcast to clients. But, for space-efficiency and semantic reasons, we do not want to retain long-term the information of how updates happened to chunked. Thus, it is sensible to store a streaming-oriented representation temporarily in Redis, in parallel to the batch-oriented representation in databases, blobs, and files.
- A Redis-based architecture would enable the websocket updates to go out before data had touched disk, reducing latency. I am not sure that low latency is very important for our use cases, but it is a nice a consequence of this design.

Lifecycle of an HTTP request being processed:

1. Data or pointers to data are received (e.g. from Bluesky) in an HTTP request.
2. The data or pointer-to-data is stashed in Redis, keyed on something like `(node_id, sequence_number)`, where the `node_id` identifies the dataset in the SQL catalog database and the `sequence_number` is a transient sequential ID generated by Redis.
3. The update, proxied from Redis pub-sub to a Tiled websocket connection, is sent to any subscribed clients.
4. The data or pointer-to-data is incorporated in the SQL database (as now).
5. When the client is done writing, it communicates this to Tiled, via some new API (`/close/{path}`?). A boolean value in the `nodes` table in the SQL catalog database representing "live or not" is updated. After some grace period (for any clients to catch up) the copy in Redis is garbage-collected.

![Image](https://github.com/user-attachments/assets/d9cd8c6e-dfca-4f1e-b8f9-fa4bb48c4123)

We propose two websockets endpoints. First, a simple endpoint that can only stream a single dataset, via one-way communication:

```
wss://.../stream/single/{path}?start_at={sequence_number}
```

where sequence number:

`N=0` - Start as far back as you have.
`N>0` - Start at a specific position (e.g after disconnection/reconnection).
`omitted` - Start at the most recent update.

Second, an endpoint that will support two-way communication, multiplexing streams over a single connection, support subscribe/unsubscribe and eventually flow control (e.g. decimation).

```
wss://.../stream/multi
```

We propose to merge the first one while researching prior art on the second one.

**Comment by danielballan** [comment]
This is implemented! Work on this is now tracked on individual issues labeled `streaming`.
